Fixed #28912 -- Fixed EmailMessage.message() to exclude an empty To field.

jdufresne opened the pull request:
https://code.djangoproject.com/ticket/28912

claudep commented on django/core/mail/message.py:
```diff
@@ -254,7 +254,8 @@ def message(self):
         msg['From'] = self.extra_headers.get('From', self.from_email)
-        msg['To'] = self.extra_headers.get('To', ', '.join(map(str, self.to)))
+        if self.to:
+            msg['To'] = self.extra_headers.get('To', ', '.join(map(str, self.to)))
```
What if `self.to` is empty and `self.extra_headers` contains a `To` header? Weird use case, but still…

jdufresne replied:
I agree.

Originally, I was following the existing pattern for [`Reply-To`](https://github.com/django/django/blob/af35c69a3c595ff77188ff0d7dac6ab747484ac3/django/core/mail/message.py#L260-L261), which has this same edge case. I wasn't sure if I should follow the pattern or fix the existing edge case.

Additionally, I noticed that [`Cc`](https://github.com/django/django/blob/af35c69a3c595ff77188ff0d7dac6ab747484ac3/django/core/mail/message.py#L258-L259) doesn't have the same fallback to `extra_headers` as the other headers. So it a sense, it too also has this edge case.

Shall I fix these other edge cases while I'm at it? I certainly don't mind.

claudep replied:
I think so. Ideally in a separate commit.

jdufresne replied:
> What if self.to is empty and self.extra_headers contains a To header? Weird use case, but still…

Turns out, this case was already handled. It is handled later during the general `.extra_headers` handling here:

https://github.com/django/django/blob/2.0/django/core/mail/message.py#L275-L278

jdufresne commented:
Applied suggested feedback:

- `Reply-To` is now set if it exists in `extra_headers` but not as an attribute
- Moved common logic into a function for reuse
- `Cc` now follows same logic
- `To` now follows same logic 

Thanks!

timmartin commented:
The change looks good to me. You'll need to squash these into a single commit before this can be merged.

jdufresne commented:
I'm happy to squash the commits if that is preferred. These were left as separate commits at the [request](https://github.com/django/django/pull/9454#discussion_r155923808) of @claudep. Please advise on what is preferred.

timgraham commented on tests/mail/tests.py:
```diff
@@ -85,6 +85,15 @@ def test_multiple_recipients(self):
+    def test_no_to_recipients(self):
+        email = EmailMessage('Subject', 'Content', 'from@example.com', cc=['cc@example.com'])
+        message = email.message()
+        self.assertEqual(message['Subject'], 'Subject')
```
Do you think checking the Subject and Message adds value? It seems to distract from the main issue that's being tested.

timgraham commented on django/core/mail/message.py:
```diff
@@ -408,6 +406,14 @@ def _create_attachment(self, filename, content, mimetype=None):
         return attachment
 
+    def _set_list_header_if_not_empty(self, msg, header, default):
+        try:
```
I think a docstring would be useful.

timgraham commented on tests/mail/tests.py:
```diff
@@ -132,6 +132,22 @@ def test_cc(self):
+            'Subject', 'Content', 'bounce@example.com', ['to@example.com'],
+            cc=['foo@example.com'], headers={'Cc': 'override@example.com'},
+        )
+        message = email.message()
```
Perhaps `.message()` could be done on the previous line after the ) since the `email` variable is otherwise unused.

timgraham commented on tests/mail/tests.py:
```diff
@@ -132,6 +132,22 @@ def test_cc(self):
+        message = email.message()
+        self.assertEqual(message['Cc'], 'override@example.com')
+
+    def test_cc_headers_only(self):
```
This doesn't seem to work as a regression test (additional test coverage is fine, but shouldn't be mixed with bug fixes as it makes it difficult to see what behavior is fixed).

timgraham commented on tests/mail/tests.py:
```diff
@@ -254,6 +254,14 @@ def test_reply_to_header(self):
         message = email.message()
         self.assertEqual(message['Reply-To'], 'override@example.com')
 
+    def test_reply_to_headers_only(self):
```
This test is passing without any changes.

timgraham commented on tests/mail/tests.py:
```diff
@@ -85,6 +85,15 @@ def test_multiple_recipients(self):
         self.assertEqual(message['From'], 'from@example.com')
         self.assertEqual(message['To'], 'to@example.com, other@example.com')
 
+    def test_no_to_recipients(self):
```
Something like `test_header_omitted_for_no_to_recipients` may be more descriptive.

jdufresne commented:
@timgraham Thanks for the review. I've made the suggested changes.

- Added a docstring to `._set_list_header_if_not_empty()` (feedback on wording welcome)
- Created a new ticket [#28971](https://code.djangoproject.com/ticket/28971) to track the Cc override issue
- Added the ticket reference to the commit message
- Moved passing tests to an earlier commit to be clear that they aren't regressions tests, but are tests for existing behavior
- Changed the test name as suggested
- Removed unrelated asserts from the tests
- Moved `.message()` to previous line in tests

I've also squashed the commit that introduces `._set_list_header_if_not_empty()` with the commit that fixes the Cc override issue. I think it better illustrates why the method was introduced. I've also tweaked some of the commit messages to better explain the current state.

All feedback welcome, ready for review. Thanks.

jdufresne commented:
This change may require an adjustment to line, https://github.com/django/django/blob/2.0/django/core/mail/message.py#L276

I'm going to investigate.

jdufresne commented:
Updated mentioned code and added a test for the incidental regression. Ready for review.

timgraham commented:
Looks good. The try/except approach seemed a bit verbose and I now I see that refactored it back to the same style as the original code. Did you find the try/except approach more readable?

jdufresne commented:
I prefer the `try`/`except` style because:

I think it slightly better follows Python's [EAFP](https://docs.python.org/3/glossary.html#term-eafp) recommended style (Easier to ask for forgiveness than permission).

Using the `try`/`except` means the `.join()` will not be called in the `.extra_headers` case. As a function argument, the `.join()` is evaluated eagerly.

But ultimately, I recognize that it is a matter of personal preference. Either works, so I'm fine with either form.
